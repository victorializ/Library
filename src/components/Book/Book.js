import React, {useState} from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import { Card, Button } from 'semantic-ui-react';

import { constants } from '../../i18n';
import { formatData } from '../../services/format-data/books';
import { order as httpOrder } from '../../services/http-client';
import { useRequest } from '../../hooks';
import { LoadingComponent } from '../LoadingComponent';
import { ErrorMessage } from '../UITable/ErrorMessage';
import { colors } from '../../assets/semantic-colors';

function Book ({item, userId}) {
    const {
        id,
        name, 
        bookYear, 
        numberAvailable,
        bookAuthors, 
        bookGenres
    } = formatData(item);
    
    const [ isOrdered, setIsOrdered ] = useState(false);
    const response = useRequest(httpOrder, [userId, id], isOrdered);


    return (
        <Card className='book'>
            <Card.Content>
                <Card.Header>
                    {name}
                </Card.Header>
                <Card.Meta>
                    {bookYear}
                </Card.Meta>
                <Card.Description>
                    {constants.available}: {numberAvailable}
                </Card.Description>
                { !!bookAuthors.length && 
                    <Card.Description className='book__authors'>
                        {constants.bookAuthors}: {bookAuthors}
                    </Card.Description>
                }
                { !!bookGenres.length && 
                    <Card.Description className='book__genres'>
                        {constants.bookGenres}: {bookGenres}
                    </Card.Description>
                }
            </Card.Content>
            {userId && !isOrdered && 
                <Card.Content extra>
                    <Button basic onClick={() => setIsOrdered(true)}>
                        Order
                    </Button>
                </Card.Content>
            }
            <LoadingComponent {...response} />
            {response.data && 
                <Button 
                    as={Link} to='/orders' 
                    color={colors.primary}>
                    {`Ordered! ${constants.personalAccount}`}
                </Button>
            }
        </Card>
    );
}

const OrderBook = connect(({user}) => user)(Book);

export { Book as BookComponent, OrderBook as Book };