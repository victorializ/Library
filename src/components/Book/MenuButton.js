import React from 'react';

import { Button } from 'semantic-ui-react'

import { constants } from '../../i18n';
import { colors } from '../../assets/semantic-colors';

const MenuButton = ({name, selectedButton, clickHandler}) => {

  const getButtonColor = (buttonName, selected) =>  {
    return buttonName === selected ? colors.primary : colors.grey;
  }

  return (
    <Button 
      onClick={() => clickHandler(name)}
      color={getButtonColor(name, selectedButton)}>
        {constants[name]}
    </Button>
  )
};

export { MenuButton };