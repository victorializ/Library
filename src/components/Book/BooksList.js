import React  from 'react';

import { Book } from './Book';
import { ItemsList } from '../hocs/ItemsList';

const BooksList = ({data}) => <ItemsList data={data} Component={Book} />

export { BooksList };    
