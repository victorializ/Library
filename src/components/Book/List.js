import React  from 'react';

import { Book } from './Book';
import { ItemsList } from '../ItemsList';

const List = ({data}) => 
    <ItemsList 
        data={data} 
        Component={Book}
    />

export { List as BooksList };    
