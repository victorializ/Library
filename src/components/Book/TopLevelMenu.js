import React  from 'react';
import { Search, Button } from 'semantic-ui-react'

import { MenuButton } from './MenuButton';
import { constants } from '../../i18n';
import { book, sorting } from '../../types';
import { colors } from '../../assets/semantic-colors';

const TopLevelMenu = ({
  filter, 
  onFilterSet, 
  input, 
  onInput, 
  sort, 
  onSortSet, 
  onSortUnsert, 
  onFilterUnsert
}) =>
  <div className="books__menu">
    <div className="books__search">
      <Button.Group>
        <Button 
          onClick={onFilterUnsert}
          color={colors.secondary}>
          {constants.filter}
        </Button>
        <MenuButton 
          name={book.name} 
          selectedButton={filter} 
          clickHandler={onFilterSet} />
        <Button.Or />
        <MenuButton 
          name={book.bookYear} 
          selectedButton={filter} 
          clickHandler={onFilterSet} />
        <Button.Or />
        <MenuButton 
          name={book.bookAuthors} 
          selectedButton={filter} 
          clickHandler={onFilterSet} />
        <Button.Or />
        <MenuButton 
          name={book.bookGenres} 
          selectedButton={filter} 
          clickHandler={onFilterSet} />
      </Button.Group>
      <Search
        value = {input}
        showNoResults={false}
        onSearchChange ={(e, {value}) => onInput(value)}
      ></Search>
    </div>
    <div className="books__filter">
      <Button.Group>
        <Button 
          onClick={onSortUnsert}
          color={colors.secondary}>
          {constants.sort}
        </Button>
        <MenuButton 
          name={sorting.asc} 
          selectedButton={sort} 
          clickHandler={onSortSet} />
        <Button.Or />
        <MenuButton 
          name={sorting.desc} 
          selectedButton={sort} 
          clickHandler={onSortSet} />
      </Button.Group>
    </div>
  </div>

export { TopLevelMenu as BooksTopLevelMenu };