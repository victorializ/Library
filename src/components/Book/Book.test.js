import React from 'react';
import { shallow } from 'enzyme';

import { BookComponent as Book } from './Book';

describe('Book component', () => {
    it('renders book component correctly', () => {
        const item = {bookAuthors:[], bookGenres:[]}
        expect(shallow(<Book item={item} />)).toMatchSnapshot();
    }); 
    it('renders book authors if available', () => {
        const item = {bookAuthors:[{author: {firstName: 'N', lastName: 'L'}}], bookGenres:[]}
        const wrapper = shallow(<Book item={item} />);
        expect(wrapper.find('.book__authors').exists()).toBeTruthy();
    }); 
    it('not renders book authors if unavailable', () => {
        const item = {bookAuthors:[], bookGenres:[]}
        const wrapper = shallow(<Book item={item} />);
        expect(wrapper.find('.book__authors').exists()).toBeFalsy();
    });
    it('renders book genres if available', () => {
        const item = {bookAuthors:[], bookGenres:[{genre: {name: 'G'}}]}
        const wrapper = shallow(<Book item={item} />);
        expect(wrapper.find('.book__genres').exists()).toBeTruthy();
    }); 
    it('not renders book genres if unavailable', () => {
        const item = {bookAuthors:[], bookGenres:[]}
        const wrapper = shallow(<Book item={item} />);
        expect(wrapper.find('.book__genres').exists()).toBeFalsy();
    });
});