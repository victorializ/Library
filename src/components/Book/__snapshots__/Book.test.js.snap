// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Book component renders book component correctly 1`] = `
<Card
  className="book"
>
  <CardContent>
    <CardHeader />
    <CardMeta />
    <CardDescription>
      Available
      : 
    </CardDescription>
  </CardContent>
  <LoadingComponent
    data={null}
    error={
      Object {
        "message": null,
      }
    }
    loading={false}
  />
</Card>
`;
